### PR TITLE
PUBDEV-5919: ARFF File Import fails if header is too large (>4MB)

### PR DESCRIPTION
**Pull request by seb-h2o**
https://0xdata.atlassian.net/browse/PUBDEV-5919

- change ARFF guess setup logic to loop through the dataset chunks as soon as the chunk looks like a valid part of the ARFF header and until the @data segment.

### CODE REVIEW

**Comment by seb-h2o** [comment]
no need for review right now: just realized that this won't work with large zipped ARFF files...

File: h2o-core/src/main/java/water/parser/ARFFParser.java [hunk 1]
```diff
@@ -14,26 +16,39 @@
...
     String[] labels;
     String[][] domains;
     String[] headerlines = new String[0];
     byte[] ctypes;
 
     // header section
     ArrayList<String> header = new ArrayList<>();
-    offset = readArffHeader(offset, header, bits, singleQuotes);
+
+    int offset;
```
[the remark this answers is not in the document]
Comment (seb-h2o) [reply h2o-core/src/main/java/water/parser/ARFFParser.java]: good catch: in `readArffHeader` method, the `line_fragment` can't be a comment, but the first bytes of the next chunk could be, making it also more difficult to detect and ignore.

File: h2o-core/src/main/java/water/parser/ARFFParser.java [hunk 2]
```diff
@@ -14,26 +16,39 @@
...
+    int chunk_idx = 0; //relies on the assumption that bits param have been extracted from first chunk: cf. ParseSetup#map
```
[the remark this answers is not in the document]
Comment (seb-h2o) [reply h2o-core/src/main/java/water/parser/ARFFParser.java]: files with long headers in archive are now handled with the H2OUnsupportedDataFileException below

**Comment by seb-h2o** [comment]
@michalkurka, @Pscheidl: thanks for looking quickly at those changes, this is needed shortly for benchmarks demo

File: h2o-core/src/main/java/water/util/ArrayUtils.java [hunk 3]
```diff
@@ -559,6 +559,11 @@ public static String arrayToString(int[] ary) {
     return false;
   }
 
+  static public boolean contains(byte[] a, byte d) {
+    for (int anA : a) if (anA == d) return true;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/ArrayUtils.java]: why do you cast to an `int`? Is that intentional?

File: h2o-core/src/main/java/water/parser/CsvParser.java [hunk 4]
```diff
@@ -60,10 +66,7 @@
     byte c = bits[offset];
     // skip comments for the first chunk (or if not a chunk)
     if( cidx == 0 ) {
-      while ( c == '#'
-              || isEOL(c)
-              || c == '@' /*also treat as comments leading '@' from ARFF format*/
-              || c == '%' /*also treat as comments leading '%' from ARFF format*/) {
+      while (ArrayUtils.contains(nonDataLineMarkers(), c) || isEOL(c)) {
```
[the remark this answers is not in the document]
Comment (seb-h2o) [reply h2o-core/src/main/java/water/parser/CsvParser.java]: as this is the main loop, better be on the safe side, so I'll call it outside the loop

File: h2o-core/src/main/java/water/util/ArrayUtils.java (hunk 3 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/util/ArrayUtils.java]: nope :) copy/pasted the int version, forgot this bit, thanks for catching this

**Comment by seb-h2o** [comment]
@michalkurka, @Pscheidl : this is ready for merging, any objection?

**Comment by michalkurka** [comment]
Yeah, please wait for Pavel's approval. @Pscheidl can you please focus on the changes in CSVParser.

Otherwise, I think the changes look good.

**Pull request merged**